Add a Hdf mixin that takes care of type info + general boilerplate

pmrv opened the pull request:
It came up a few times that `DataContainer` was subclassed because of the HDF functionality, but wasn't otherwise very suitable, e.g. [here](https://github.com/pyiron/pyiron_contrib/pull/205) and [here](https://github.com/pyiron/pyiron_contrib/pull/228) and generally we have objects that write themselves to HDF but are not based on a general data structure like `DataContainer`.  I've had this sketch for a while now that should make it easier to implement this kind of classes.

As a demonstration I've derived `DataContainer` from this new mixin, but generally it should be easy to move old classes to this format (expect `GenericParameters` which handles HDF in a slightly different way).

Test + docs will follow after discussion whether we find this useful.

pmrv committed:
First stab at HDF mixin

pmrv committed:
Turn WithHDF into a slots object

pmrv committed:
Add _type_to_hdf to interface

pmrv committed:
Implement _type_to_hdf

pmrv committed:
Pass HDF version to _from_hdf

pmrv committed:
Move HasHDF to pyiron_base.interfaces

Adds a method to allow sub classes to specify an HDF group name that
they want to be written into.  Will be used if to_hdf() is not given a
group_name.

Pass self in abstractmethods

pmrv committed:
Rename get_group_name method

pmrv committed:
Check if HDF group is empty before writing

pmrv committed:
Use _get_hdf_group_name in to_hdf

pmrv committed:
Add HasHDF to DataContainer

pmrv committed:
Keep old OBJECT field in type information

It's unused in all our code, but removing should involve discussion

pmrv committed:
Save __version__ only if defined

pmrv committed:
Implement rewrite_hdf

pmrv committed:
Write type info after calling _to_hdf

This is so the objects implementation cannot muck it up accidentally.

pmrv committed:
Fix import

niklassiemer commented:
I really like this idea a lot! It might make writing new self-hdf-able classes a lot easier! Maybe you should add the `@classmethod` `from_hdf_args(cls, hdf)` somehow, since this one allows to pass in needed `__init__` variables upon `to_object()`.

liamhuber commented on pyiron_base/interfaces/has_hdf.py:
```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
+
+class WithHDF:
```
```suggestion
class _WithHDF:
```

Or?

liamhuber commented on pyiron_base/interfaces/has_hdf.py:
```diff
@@ -0,0 +1,63 @@
+        pass
+
+    @abstractmethod
+    def _get_hdf_group_name(self):
```
```suggestion
    @property
    @abstractmethod
    def _hdf_group_name(self):
```

At least I think that's the right order for the decorators. 

Am I missing a side effect?

pmrv replied:
It should work, I didn't do it before because it's a private method anyway, but I can change it now.

liamhuber reviewed (approved):
I love this. `WithHDF` took me a second read-through to understand, but I find the use of `__enter__` to handle the group name absolutely wonderful.

I posted a couple of minor suggestions to consider.

pmrv committed:
Make WithHDF private

pmrv commented on pyiron_base/interfaces/has_hdf.py:
```diff
@@ -0,0 +1,63 @@
+        hdf["HDF_VERSION"] = self.__hdf_version__
+
+    def from_hdf(self, hdf, group_name=None):
+        group_name = group_name or self._get_hdf_group_name()
```
Explicit check for `None`.

pmrv commented on pyiron_base/interfaces/has_hdf.py:
```diff
@@ -0,0 +1,63 @@
+            self._to_hdf(hdf)
+            self._type_to_hdf(hdf)
+
+    def rewrite_hdf(self, hdf, group_name=None):
```
Garbage collection for HDF?

pmrv committed:
Add from_hdf_args to Mixin

pmrv committed:
Add copyright blurb

pmrv committed:
Don't define version by default

Some objects only have an HDF version unlike jobs that also have a code
version.

pmrv committed:
Explicitly check for None

pmrv committed:
Copy hdf object in HDFStubs

Not copying caused an error in lazy DataContainers, because with the new
HasHDF inheritance it would:
1. hdf.open(...) and pass that to DataContainer._from_hdf
2. this method would then pass references to the HDF object to the
   HDFStub
3. HasHDF.from_hdf then calls hdf.close(), this modifies the hdf object
   passed to DataContainer in place, changing where it points to in the
   file
4. At some point HDFStub.load() would be called, but since the place
   where hdf points to changed under its feet HDFStub.load() then
   reaches into the wrong location in the file

pmrv commented:
Fixed the test, details in the commits.  The problem was that previously `DataContainer.from_hdf` didn't call `hdf.close()` if it descended into a group.  `HasHDF` does that now, but since `DataContainer` passes a reference to `HDFStub`, this in turn reached into the wrong location in the file (since where hdf points to changes on `hdf.close`).

pmrv commented:
Rebased to resolve merge conflict.

liamhuber commented:
Beautiful, tests are passing so I'm happy! If you have no further concerns, please merge it and let's get this thing live :)

Since `PyironObject`s are the trifecta of python, database, and file storage, we should consider if we want some sort of `class Storable(HasHDF, HasStorage)` as an intermediate that never needs to be located in the database.

EDIT: spelling of trifecta

pmrv committed:
Add type annotations

pmrv committed:
Add docstring and usage example

pmrv committed:
More docstrings

pmrv committed:
Talk about version

pmrv committed:
Run doctests

pmrv commented:
> Beautiful, tests are passing so I'm happy! If you have no further concerns, please merge it and let's get this thing live :)
> 
> Since `PyironObject`s are the trifecta of python, database, and file storage, we should consider if we want some sort of `class Storable(HasHDF, HasStorage)` as an intermediate that never needs to be located in the database.
> 
> EDIT: spelling of trifecta

Since `HasStorage` just has the input/output container, I'd actually think, that it should just derive from `HasHDF`.

liamhuber commented:
> Since HasStorage just has the input/output container, I'd actually think, that it should just derive from HasHDF.

Yes, this can work. We just need to think a little bit whether (1) `HasStorage._to_hdf` calls `self._storage.to_hdf()`, or if (2) `HasStorage` overrides `to_hdf` with a `super` call and the storage saving.

(1) Pro: the ABC is satisfied and people can directly make new classes without ever worrying about storage (as long as everything they want serialized gets slapped onto the `storage` attribute. Con: If someone *does* want to save extra stuff they need to remember to call `super` in `_to_hdf`.

(2) Pro: `_to_hdf` maintains its status as "that thing I should do my own work in". Con: anyone inheriting from `HasStorage` is still stuck implementing `_to/_from_hdf` every single time, even if they don't need it.

I lean towards (1). Thoughts?

pmrv commented:
Yeah, (1) sounds cleaner.  Overloading the saving stuff should only be necessary for fundamental object and jobs anyway.
